### PR TITLE
3D odometry filtering

### LINKED ISSUE #5: 3D sensor + motion models
Implement 3D sensors and motion models plugins

**motion models**
- [x] unicycle_3d --> move to omnidirectional3d

**sensor models**
- [x] odometry_3d
- [x] imu_3d

### PR DESCRIPTION
**Pull request by roncapat**
Track:
- https://github.com/giafranchini/fuse/issues/1
- https://github.com/giafranchini/fuse/issues/5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # fuse
 
-**NOTE**: The `rolling` branch is a [work in progress](https://github.com/locusrobotics/fuse/issues/276) port of the fuse stack to ROS 2, it is **not** expected to work until said work is done!
+**NOTE**: The `humble` branch is a [work in progress](https://github.com/locusrobotics/fuse/issues/276) port of the fuse stack to ROS 2, it is **not** expected to work until said work is done!
```
Comment (roncapat) [inline README.md]:
```suggestion
**NOTE**: The `rolling` branch is a [work in progress](https://github.com/locusrobotics/fuse/issues/276) port of the fuse stack to ROS 2, it is **not** expected to work until said work is done!
```

Target branch is `rolling`. Please revert this change.

File: fuse_constraints/fuse_plugins.xml [hunk 2]
```diff
@@ -5,12 +5,24 @@
     the 2D angular acceleration.
     </description>
   </class>
+  <class type="fuse_constraints::AbsoluteAccelerationAngular3DStampedConstraint" base_class_type="fuse_core::Constraint">
+    <description>
+    A constraint that represents either prior information about a 3D angular acceleration, or a direct measurement of
+    the 2D angular acceleration.
```
Comment (roncapat) [inline fuse_constraints/fuse_plugins.xml]:
```suggestion
    the 3D angular acceleration.
```
is this a typo?

File: fuse_constraints/fuse_plugins.xml [hunk 3]
```diff
@@ -5,12 +5,24 @@
...
   <class type="fuse_constraints::AbsoluteAccelerationLinear2DStampedConstraint" base_class_type="fuse_core::Constraint">
     <description>
     A constraint that represents either prior information about a 2D linear acceleration, or a direct measurement of
     the 2D linear acceleration.
     </description>
   </class>
+  <class type="fuse_constraints::AbsoluteAccelerationLinear3DStampedConstraint" base_class_type="fuse_core::Constraint">
+    <description>
+    A constraint that represents either prior information about a 3D linear acceleration, or a direct measurement of
+    the 2D linear acceleration.
```
Comment (roncapat) [inline fuse_constraints/fuse_plugins.xml]:
```suggestion
    the 3D linear acceleration.
```
is this a typo?

File: fuse_constraints/fuse_plugins.xml [hunk 4]
```diff
@@ -35,12 +47,24 @@
     the 2D angular velocity.
     </description>
   </class>
+  <class type="fuse_constraints::AbsoluteVelocityAngular3DStampedConstraint" base_class_type="fuse_core::Constraint">
+    <description>
+    A constraint that represents either prior information about a 3D angular velocity, or a direct measurement of
+    the 2D angular velocity.
```
Comment (roncapat) [inline fuse_constraints/fuse_plugins.xml]:
```suggestion
    the 3D angular velocity.
```
is this a typo?

File: fuse_constraints/fuse_plugins.xml [hunk 5]
```diff
@@ -35,12 +47,24 @@
...
   <class type="fuse_constraints::AbsoluteVelocityLinear2DStampedConstraint" base_class_type="fuse_core::Constraint">
     <description>
     A constraint that represents either prior information about a 2D linear velocity, or a direct measurement of
     the 2D linear velocity.
     </description>
   </class>
+  <class type="fuse_constraints::AbsoluteVelocityLinear3DStampedConstraint" base_class_type="fuse_core::Constraint">
+    <description>
+    A constraint that represents either prior information about a 3D linear velocity, or a direct measurement of
+    the 2D linear velocity.
```
Comment (roncapat) [inline fuse_constraints/fuse_plugins.xml]:
```suggestion
    the 3D linear velocity.
```
is this a typo?

File: fuse_constraints/fuse_plugins.xml [hunk 6]
```diff
@@ -60,7 +84,7 @@
   </class>
   <class type="fuse_constraints::AbsolutePose3DStampedConstraint" base_class_type="fuse_core::Constraint">
     <description>
-    A constraint that represents either prior information about a 3D pose, or a direct measurement of the 3D pose.
+    A constraint that represents either prior information about a 3D pose, or a direct measurement of the 3D pose in quat.
```
Comment (roncapat) [inline fuse_constraints/fuse_plugins.xml]:
```suggestion
    A constraint that represents either prior information about a 3D pose, or a direct measurement of the 3D pose.
```
is this needed?

File: fuse_constraints/include/fuse_constraints/absolute_constraint.hpp [hunk 7]
```diff
@@ -46,11 +46,16 @@
 #include <fuse_core/serialization.hpp>
 #include <fuse_core/uuid.hpp>
 #include <fuse_variables/acceleration_angular_2d_stamped.hpp>
+#include <fuse_variables/acceleration_angular_3d_stamped.hpp>
 #include <fuse_variables/acceleration_linear_2d_stamped.hpp>
+#include <fuse_variables/acceleration_linear_3d_stamped.hpp>
 #include <fuse_variables/orientation_2d_stamped.hpp>
+// #include <fuse_variables/orientation_3d_stamped.hpp>
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/absolute_constraint.hpp]: Why commented?

File: fuse_constraints/include/fuse_constraints/normal_delta_orientation_3d_cost_functor.hpp [hunk 8]
```diff
@@ -120,7 +120,9 @@ class NormalDeltaOrientation3DCostFunctor
 
     // Scale the residuals by the square root information matrix to account for the measurement
     // uncertainty.
-    Eigen::Map<Eigen::Matrix<T, Eigen::Dynamic, 1>> residuals_map(residuals, A_.rows());
+    // TODO: Probably here we can set 3 as row dimension instead of Eigen::Dynamic
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_delta_orientation_3d_cost_functor.hpp]: Fix TODO

File: fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp [hunk 9]
```diff
@@ -83,12 +83,12 @@ class NormalPriorOrientation3DEulerCostFunctor
    *
    * @param[in] A The residual weighting matrix, most likely the square root information matrix. Its
    *              order must match the values in \p axes.
-   * @param[in] b The orientation measurement or prior. Its order must match the values in \p axes.
+   * @param[in] b The orientation measurement or prior.
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp]: Why this comment is removed? No more valid?

File: fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp [hunk 10]
```diff
@@ -133,19 +153,22 @@ class NormalPriorOrientation3DEulerCostFunctor
...
+
+    Eigen::Map<Eigen::Matrix<T, 3, 1>> residuals_map(residuals, A_.rows());
     residuals_map.applyOnTheLeft(A_.template cast<T>());
 
     return true;
   }
 
 private:
-  fuse_core::MatrixXd A_;  //!< The residual weighting matrix, most likely the square root
+  fuse_core::Matrix3d A_;  //!< The residual weighting matrix, most likely the square root
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp]: Switching to static size here is needed? Is this a standalone optimization that can also be applied upstream?

File: fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp [hunk 11]
```diff
@@ -133,19 +153,22 @@ class NormalPriorOrientation3DEulerCostFunctor
...
                            //!< information matrix
-  fuse_core::VectorXd b_;  //!< The measured 3D orientation (quaternion) value
+  fuse_core::Vector4d b_;  //!< The measured 3D orientation (quaternion) value
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp]: Switching to static size here is needed? Is this a standalone optimization that can also be applied upstream?

File: fuse_constraints/include/fuse_constraints/normal_prior_pose_3d.hpp [hunk 12]
```diff
@@ -0,0 +1,106 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, Locus Robotics
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_prior_pose_3d.hpp:4]:
```suggestion
 *  Copyright (c) 2020, Locus Robotics
 *  Copyright (c) 2023, Giacomo Franchini
```

File: fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp [hunk 13]
```diff
@@ -79,7 +80,7 @@ class NormalPriorPose3DCostFunctor
    *              order (x, y, z, qx, qy, qz)
    * @param[in] b The 3D pose measurement or prior in order (x, y, z, qw, qx, qy, qz)
    */
-  NormalPriorPose3DCostFunctor(const fuse_core::Matrix6d & A, const fuse_core::Vector7d & b);
+  NormalPriorPose3DCostFunctor(const fuse_core::MatrixXd & A, const fuse_core::Vector7d & b);
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp]: Why instead here we move to a dynamic size matrix?

File: fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp [hunk 14]
```diff
@@ -88,39 +89,42 @@ class NormalPriorPose3DCostFunctor
   bool operator()(const T * const position, const T * const orientation, T * residual) const;
 
 private:
-  fuse_core::Matrix6d A_;
+  fuse_core::MatrixXd A_;
```
Comment (roncapat) [inline fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp]: Why instead here we move to a dynamic size matrix?

File: fuse_constraints/src/absolute_pose_3d_stamped_constraint.cpp [hunk 15]
```diff
@@ -64,7 +116,7 @@ void AbsolutePose3DStampedConstraint::print(std::ostream & stream) const
          << "  position variable: " << variables().at(0) << "\n"
          << "  orientation variable: " << variables().at(1) << "\n"
          << "  mean: " << mean().transpose() << "\n"
-         << "  sqrt_info: " << sqrtInformation() << "\n";
+         << "  sqrt_info: \n" << sqrtInformation() << "\n";
```
Comment (roncapat) [inline fuse_constraints/src/absolute_pose_3d_stamped_constraint.cpp]:
```suggestion
         << "  sqrt_info: " << sqrtInformation() << "\n";
```
better to revert to reduce diff size.

File: fuse_constraints/src/normal_prior_pose_3d.cpp [hunk 16]
```diff
@@ -0,0 +1,157 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, Clearpath Robotics
```
Comment (roncapat) [inline fuse_constraints/src/normal_prior_pose_3d.cpp]:
```suggestion
 *  Copyright (c) 2020, Clearpath Robotics
 *  Copyright (c) 2023, Giacomo Franchini
```

File: fuse_models/include/fuse_models/common/sensor_config.hpp [hunk 17]
```diff
@@ -112,6 +138,91 @@ std::enable_if_t<is_angular_2d<T>::value, size_t> toIndex(const std::string & di
...
+//   }
+//   if (lower_dim == "yaw" || lower_dim == "z") {
+//     return static_cast<size_t>(fuse_variables::Orientation3DStamped::Euler::YAW) - 3UL;
+//   }
+
+//   throwDimensionError(dimension);
+
+//   return 0u;
+// }
+
```
Comment (roncapat) [inline fuse_models/include/fuse_models/common/sensor_config.hpp]: Can be removed or somehow useful in future?

**Review by roncapat — COMMENTED** [review]
@giafranchini Some trivial comments while I was giving a look :)

File: fuse_constraints/include/fuse_constraints/absolute_constraint.hpp (hunk 7 above)
Comment (giafranchini) [reply fuse_constraints/include/fuse_constraints/absolute_constraint.hpp]: The absolute constraint template class is actually used only for velocity and acceleration constraints. Pose constraints (and so orientation ones), are implemented in different dedicated classes. Still, we could add it here a templated version of orientation 3D for completion.

File: fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp (hunk 9 above)
Comment (giafranchini) [reply fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp]: My changes in this file came from tests that I tried some time ago and are not valid. I will revert it to the previous commit

File: fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp (hunk 10 above)
Comment (giafranchini) [reply fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp]: My changes in this file came from tests that I tried some time ago and are not valid. I will revert it to the previous commit

File: fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp (hunk 11 above)
Comment (giafranchini) [reply fuse_constraints/include/fuse_constraints/normal_prior_orientation_3d_euler_cost_functor.hpp]: My changes in this file came from tests that I tried some time ago and are not valid. I will revert it to the previous commit

File: fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp (hunk 13 above)
Comment (giafranchini) [reply fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp]: Dynamic size here is needed in order to accept partial poses relative to partial measurements. Probably this could have been implemented using directly constraints classes with orientation in RPY, instead of reconverting back in quaternion. We should discuss about this

File: fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp (hunk 14 above)
Comment (giafranchini) [reply fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.hpp]: Same as above

File: fuse_models/include/fuse_models/common/sensor_config.hpp (hunk 17 above)
Comment (giafranchini) [reply fuse_models/include/fuse_models/common/sensor_config.hpp]: can be removed

File: fuse_optimizers/include/fuse_optimizers/optimizer.hpp
Comment (roncapat) [inline fuse_optimizers/include/fuse_optimizers/optimizer.hpp:1]: Can be reverted to reduce diff size :)

File: fuse_models/test/test_sensor_proc.cpp
Comment (roncapat) [inline fuse_models/test/test_sensor_proc.cpp:1]: Can be reverted to reduce diff size :)

File: fuse_optimizers/src/fixed_lag_smoother.cpp
Comment (roncapat) [inline fuse_optimizers/src/fixed_lag_smoother.cpp:1]: Can be reverted to reduce diff size :)

File: fuse_tutorials/launch/range_sensor_tutorial.launch.py
Comment (roncapat) [inline fuse_tutorials/launch/range_sensor_tutorial.launch.py:1]: Can be reverted to reduce diff size :)

**Review by giafranchini — APPROVED** [review]

**Comment by giafranchini** [comment]
@roncapat I would like to change the branch name before opening the PR to fuse official repo, should I do that and close this PR or it's better to just merge it into rolling and then open the PR directly from rolling?

**Comment by roncapat** [comment]
It depends, branches can also be renamed if you want :) Just go to the repository homepage, click "branches"and click the pencil icon. You may prefer to then open a PR towards upstream keeping all the history (commits in this branch) or squash something/everything.
